### PR TITLE
Components: use new theming accent color in all components

### LINKED ISSUE #45249: Components theming: use new `accent` color in every component
Part of #44116

## What

Swap hardcoded accent/theme color styles in favour of the newly introduced CSS variables, both for Sass and Emotion-based components.

There will be [three variables to refactor towards](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/utils/theme-variables.scss):
- `$components-color-accent`
- `$components-color-accent-darker-10`
- `$components-color-accent-darker-20`

## Why

Part of the #44116 effort to add theming to the Components library.

## How

Mainly a search and replace process, work is being performed in #45289.

### PR DESCRIPTION
**Pull request by chad1008**
closes #45249
part of #44116

## What?
Updates the Components package to use the new accent color variables in all relevant components.

## Why
Because colors look nice.

## How? 
First updating the emotion variables to use the new component theme var, with the proper fallbacks.

Then updating any references to the existing WP theme variables to instead point to the new [component theme variables](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/utils/theme-variables.scss) in individual component styles.

**TODO List**
- [x] Update emotion variables
- [x] AnglePickerControl
- [x] Autocomplete
- [x] useBaseField hook
- [x] CheckBoxControl
- [x] ColorPalette
- [x] DropZone
- [x] FormToggle
- [x] FormTokenField
- [x] useNavigateRegions
- [x] MenuItem
- [x] Notice
- [x] Panel
- [x] RangeControl
- [x] ResizableBox
- [x] SearchControl
- [x] Snackbox
- [x] Spinner
- [x] TabPanel
- [x] ToggleGroupControl
- [x] ToolsPanel
- [x] Input (variable is in `utils/input/base.js`)

For each of the above, we'll need to:
- update the variables
- test the default and theme colors in Storybook
- test the default an theme colors in the editor:
  - run GB locally and load the editor to a state that will render the desired component
  - modify the desired component by wrapping its render value in a `<Theme>` component
  - pass the `Theme` component different `accent` props to test the color change

## Testing Instructions
- Launch Storybook locally
- Play with as many components as you can/want
- If applicable, confirm that the default color displays as expected.
- If applicable, test interactions (like hovering) to ensure the darker color variants are working
- If applicable, use the paintbrush decorator icon in the top toolbar to test different themes. Confirm the accent color changes when a new theme is selected. 

## Additional Notes

Some things that came up, but don't necessarily fit in this PR. Some of these are mentioned in comments below, but I wanted to gather them in a single location:

- [Outline visible](https://github.com/WordPress/gutenberg/pull/45289#issuecomment-1291819631) while rotating the `AngleCircle` of `AnglePickerControl`. This outline is immune to theming, but we should be able to fix that easily. Question is do we want that outline showing? Or do we want to hide it?
- Certain colors [could potentially cause some confusion with `Notice` styling](https://github.com/WordPress/gutenberg/pull/45289#pullrequestreview-1157189100). We may need to update `Notice's` Readme.
- There are several places where [styles from the `base-styles` package need to be overridden for theming to take effect](https://github.com/WordPress/gutenberg/pull/45289#discussion_r1005976003). On this PR, all such overrides are commented and separated into their own commits in case we decide not to override them.
- `COLORS.ui.borderFocused` is mapped to the `darker-10` color variant and as of this PR, our new `$components-color-accent-darker-10` variable. That `borderFocused` property [isn't used exclusively for borders though](https://github.com/WordPress/gutenberg/pull/45289/files/6f0d8ee3715bb1ce68ace024252ef0c6e6f77204#diff-b9d2f04e5531eddd009a262929fd75fa8cf52cd4a1b98a42a29fa9a863ac8ba5). Should it be renamed?
- the `Inserter` (a `block-library` component) has SVGs icons for the various blocks that do not respect our theming. I don't think a simple CSS override will do here? 🤔 
- `block-editor`'s `URLPopover`, `TextDecorationControl`, and `TextTransformControl` don't currently respect theming. We can probably override those styles without too much trouble, but I didn't want to hold this PR up while I dove down that rabbit hole. If desired, I can happily work to add those changes here 🙂

### CODE REVIEW

**Comment by chad1008** [comment]
Leaving this as a note for Future Me:

On `AnglePickerControl` a small interior outline is visible as the user rotates the `AngleCircle`. This outline is currently controlled by browser presets:

Screen Shot 2022-10-26 at 06 00 11

```
:focus-visible {
    outline: -webkit-focus-ring-color auto 1px;
}
```

With the original default blue this didn't really stand out, and even looked potentially intentional. With themes showing a different color though, it's a bit jarring.

We'll probably want a followup to `CircleIndicatorWrapper` in `angle-picker-control-styles.js` to either remove that outline in this case, or at least match the color.

File: packages/components/src/checkbox-control/style.scss [hunk 1]
```diff
@@ -27,16 +27,17 @@ $checkbox-input-size-sm: 24px; // Width & height for small viewports.
 	@include reduce-motion("transition");
 
 	&:focus {
-		box-shadow: 0 0 0 ($border-width * 2) $white, 0 0 0 ($border-width * 2 + $border-width-focus) var(--wp-admin-theme-color);
+		box-shadow: 0 0 0 ($border-width * 2) $white, 0 0 0 ($border-width * 2 + $border-width-focus) $components-color-accent;
+		border-color: $components-color-accent;
```
Comment (chad1008) [inline packages/components/src/checkbox-control/style.scss]: `border-color` was added because without it, the focused input box retained the default `--wp-admin-theme-color` border color:

<table>
<tr>
	<td>Before:</td>
	<td>After:</td>
</tr>
<tr>
	<td>
		Screen Shot 2022-10-26 at 07 43 30
	</td>
	<td>
		Screen Shot 2022-10-26 at 07 43 09
	</td>
</tr>
</table>

I wasn't able to identify the source of the clashing style, so it's either something coming from outside the package, or just something more experienced eyes will be able to locate more easily 🙂

Comment (chad1008) [reply packages/components/src/checkbox-control/style.scss]: On second thought, as I continue working this doesn't feel like the right solution. I'm seeing other, similar border issues (e.g. on `FormTokenField`).

They look suspiciously like the `focus` styles in `utils/input/base.js` (which will be updated during the course of this PR), but updating the vars in that file doesn't seem to have an impact. Once I sort out what's actually going on, this added `border-color` declaration will probably be removed.

Comment (chad1008) [reply packages/components/src/checkbox-control/style.scss]: After some guidance from @ciampo, was able to identify the `@wordpress/base-styles mixins` that were at the root of these colors. I've also added `&[aria-checked="mixed"]` to the CheckboxControl's styles, as that was also using the `--wp-admin-theme-color` in the mixin.

Comment (ciampo) [reply packages/components/src/checkbox-control/style.scss]: @mirka this may be a recurring theme (pun not intended) — some styles may come from the `base-styles` package, and we'll have to decide if we should override them in the components package with our own theme variable, or if we should just leave them as they are.

File: packages/components/src/form-token-field/style.scss [hunk 2]
```diff
@@ -12,6 +12,13 @@
 
 	&.is-active {
 		@include input-style__focus();
+		border-color: $components-color-accent;
+		box-shadow: 0 0 0 ($border-width-focus - $border-width) $components-color-accent;
+	}
+
+	&:focus {
+		border-color: $components-color-accent;
+		box-shadow: 0 0 0 ($border-width-focus - $border-width) $components-color-accent;
```
Comment (chad1008) [inline packages/components/src/form-token-field/style.scss]: `border-color` and `box-shadow` were previously defined by `@wordpress/base-styles` for both `.is-active` and `:focus`. For now, overrides have been added to the component styles to allow theming to take effect.

File: packages/components/src/notice/style.scss [hunk 3]
```diff
@@ -3,7 +3,7 @@
 	font-family: $default-font;
 	font-size: $default-font-size;
 	background-color: $white;
-	border-left: 4px solid var(--wp-admin-theme-color);
+	border-left: 4px solid $components-color-accent;
```
Comment (chad1008) [inline packages/components/src/notice/style.scss:6]: This is an interesting case for theming... on some themes (like the "Sunset" theme we're currently using in Storybook) the notices can take on an orange color. With another theme it could even be red. These colors on notices usually have more meaning than they would in other components, indicating the severity/urgency of the notice.

Not sure if/how that impacts how we'd want to apply theming on this component.

File: packages/components/src/search-control/style.scss [hunk 4]
```diff
@@ -18,7 +18,8 @@
 
 		&:focus {
 			background: $white;
-			box-shadow: inset 0 0 0 var(--wp-admin-border-width-focus) var(--wp-admin-theme-color);
+			box-shadow: inset 0 0 0 var(--wp-admin-border-width-focus) $components-color-accent;
+			border-color: $components-color-accent;
```
Comment (chad1008) [inline packages/components/src/search-control/style.scss]: `border-color` was previously defined by `@wordpress/base-styles` for`:focus`. For now, an override has been added to the component styles to allow theming to take effect.

File: packages/components/src/notice/style.scss (hunk 3 above)
Comment (ciampo) [reply packages/components/src/notice/style.scss:6]: That's an interesting point. I will note that the `success`, `warning` and `error` statues of notice already have a hardcoded border color (a few lines below in this file), and so this specific rule would only affect the "default" and 
"informational" statuses.

Current behaviour is also explained in [the README](https://github.com/WordPress/gutenberg/blob/9efab2cbe695ff831c0c8b099b3193629c4989a7/packages/components/src/notice/README.md#usage) — we should keep that in mind in case it needs updating too

File: packages/components/src/tools-panel/styles.ts [hunk 5]
```diff
@@ -148,7 +148,7 @@ export const DropdownMenu = css`
 `;
 
 export const ResetLabel = styled.span`
-	color: var( --wp-admin-theme-color-darker-10 );
+	color: ${ COLORS.ui.borderFocus };
```
Comment (chad1008) [inline packages/components/src/tools-panel/styles.ts]: Along with this change, there are a couple of other pre-existing places in the codebase where this variable is being used for things other than a border... do we want to consider renaming it in `utils/color-values.js` in a followup PR?

File: packages/components/src/checkbox-control/style.scss (hunk 1 above)
Comment (chad1008) [reply packages/components/src/checkbox-control/style.scss]: There were a few more of these as well. I've added comments and/or isolated those changes in separate commits so they're easy to remove if we decide to leave them alone for the time being.

**Comment by chad1008** [comment]
Looking at the current unit test failure, I think we need to update a[ snapshot in the `edit-post` package](https://github.com/WordPress/gutenberg/blob/dd69919b8f2bf1b11fca5f88996f7bf5051c9d1a/packages/edit-post/src/components/preferences-modal/test/__snapshots__/index.js.snap#L582-L591) to account for these new variables. Am I right in thinking it makes sense to do that in this same PR so the changes and the test remain in sync?

File: packages/components/src/notice/style.scss (hunk 3 above)
Comment (chad1008) [reply packages/components/src/notice/style.scss:6]: > Current behaviour is also explained in [the README](https://github.com/WordPress/gutenberg/blob/9efab2cbe695ff831c0c8b099b3193629c4989a7/packages/components/src/notice/README.md#usage) — we should keep that in mind in case it needs updating too

Good call. I've added an update to the README ✅

**Comment by ciampo** [comment]
> Looking at the current unit test failure, I think we need to update a[ snapshot in the edit-post package](https://github.com/WordPress/gutenberg/blob/dd69919b8f2bf1b11fca5f88996f7bf5051c9d1a/packages/edit-post/src/components/preferences-modal/test/__snapshots__/index.js.snap#L582-L591) to account for these new variables. Am I right in thinking it makes sense to do that in this same PR so the changes and the test remain in sync?

Yep, let's update snapshots as part of this PR

**Comment by chad1008** [comment]
> Yep, let's update snapshots as part of this PR

Done. Rebased the PR, waiting to see what CI thinks of it now!

File: packages/components/src/tools-panel/styles.ts (hunk 5 above)
Comment (mirka) [reply packages/components/src/tools-panel/styles.ts]: We haven't yet audited these "semantic" color variables yet, so I don't know whether any of them are still relevant. The semantics are potentially useful when theming so we should keep them around until we audit them. That also means I don't want to force any new usages into these existing semantics if it doesn't make sense.

This ResetLabel isn't related to a `borderFocus`, so I would consider exposing a `COLORS.ui.themeDark10` that we can use here. (Aaand it looks like you already did that! 😄)

Also unclear why this ResetLabel is a darker-10 and not just the theme color 🤔 Do you remember, @aaronrobertshaw?

File: packages/components/src/button/style.scss [hunk 6]
```diff
@@ -324,6 +324,11 @@
 	.components-visually-hidden {
 		height: auto;
 	}
+
+	// Override for theming on the block mover handle. `.block-editor-block-mover__drag-handle` was needed for specificity to override declaration order.
+	.components-accessible-toolbar &.block-editor-block-mover__drag-handle:focus::before {
+		box-shadow: inset 0 0 0 var(--wp-admin-border-width-focus) $components-color-accent, inset 0 0 0 4px $white;
+	}
```
Comment (mirka) [inline packages/components/src/button/style.scss]: I am very reluctant to let `block-editor` taint this file again because Marco just recently cleaned it up 😄

I'm thinking we should leave all the non-components-package stuff for later so we can address it in a coordinated way.

File: packages/components/src/checkbox-control/style.scss (hunk 1 above)
Comment (mirka) [reply packages/components/src/checkbox-control/style.scss]: Interesting! I played around with how CSS variable reassignment works to see if it would suffice for our use case, but unfortunately it all becomes circular so it won't work 😂

```scss
.component-foo {
  // It would've been nice if the var() was evaluated before assignment, but it doesn't work like that 🙃
  --wp-admin-theme-color: var(--wp-admin-theme-color, somefallback);

  @include some-base-styles-mixin;
}
```

I am actually now considering decoupling wp-components from some of the base-styles mixins (i.e. copy/paste the relevant stuff over). The mixin dependencies were already hard to work with in wp-components, not to mention being a blocker for Emotion migration. In this CheckboxControl for example, you can see some of the styles are duplicated for some/no reason. And the fact that WP's "canonical" CheckboxControl styles have already diverged from the base-styles mixin means that the mixin isn't useful for style consistency anyway 🤷

tl;dr — For this PR I am fine with leaving the base-styles overrides out, so we can address them in a coordinated way later.

**Review by mirka — COMMENTED** [review]
This is great, thank you. It's super helpful how you exposed a bunch of potential problems and documented them so well.

File: packages/components/src/checkbox-control/style.scss (hunk 1 above)
Comment (ciampo) [reply packages/components/src/checkbox-control/style.scss]: > I am actually now considering decoupling wp-components from some of the base-styles mixins

That's my gut feeling too. On one hand, those mixins have the potential to provide consistent styles throughout the whole Gutenberg repo — but on the other hand, they are written in a way that makes them too leaky and hard to work with as a "loose" dependency.

If the net result is that today we have to write many overrides for those mixins anyway, I agree that we should decouple them from wpcomponents, and move over the relevant stuff in a way that works better for the package

Comment (ciampo) [reply packages/components/src/checkbox-control/style.scss]: > tl;dr — For this PR I am fine with leaving the base-styles overrides out, so we can address them in a coordinated way later.

Should we have a tracking issue where we list the overrides? Or do you think that the problem is so widespread that it's a wasted effort?

Comment (mirka) [reply packages/components/src/checkbox-control/style.scss]: > Should we have a tracking issue where we list the overrides? Or do you think that the problem is so widespread that it's a wasted effort?

If it's useful for task coordination, sure. I think it's mostly just `@include input-control`.

File: packages/components/src/button/style.scss (hunk 6 above)
Comment (ciampo) [reply packages/components/src/button/style.scss]: > I'm thinking we should leave all the non-components-package stuff for later so we can address it in a coordinated way.

Do you mean all overrides, including the ones related to the base styles? Or just the overrides needed for other packages too (i.e block editor) ?

File: packages/components/src/checkbox-control/style.scss (hunk 1 above)
Comment (ciampo) [reply packages/components/src/checkbox-control/style.scss]: Cool. I also proposed it since Chad mentioned that there are a few instances where he had to add such overrides:

> I've added comments and/or isolated those changes in separate commits so they're easy to remove if we decide to leave them alone for the time being.

File: packages/components/src/button/style.scss (hunk 6 above)
Comment (mirka) [reply packages/components/src/button/style.scss]: In this comment I was referring to the latter (other packages like block-editor). But yeah, I think base-styles is better addressed separately as well.

The basic principle I'm trying to stick with is that `Theme` is a wp-components concept, and wp-components is the upstream dependency, so if downstream packages like block-editor want to participate they need to use whatever API that wp-components has provided. What that API should look like is to be decided, but we'll probably need to export `$components-color-accent`, `COLORS.ui.theme`, etc from wp-components.

Comment (ciampo) [reply packages/components/src/button/style.scss]: > In this comment I was referring to the latter (other packages like block-editor). But yeah, I think base-styles is better addressed separately as well.
>
> The basic principle I'm trying to stick with is that Theme is a wp-components concept, and wp-components is the upstream dependency, so if downstream packages like block-editor want to participate they need to use whatever API that wp-components has provided.

Agreed on both points.

> What that API should look like is to be decided, but we'll probably need to export `$components-color-accent`, `COLORS.ui.theme`, etc from wp-components.

We'll have to be careful in doing this, since we won't be able to undo these API decisions (because of WordPress backwards compat rules).

I also wonder what is the better format to expose those theme variables — as Sass + Emotion vars, as CSS vars.

File: packages/components/src/tools-panel/styles.ts (hunk 5 above)
Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/styles.ts]: > Also unclear why this ResetLabel is a darker-10 and not just the theme color 🤔 Do you remember

Thanks for the ping. 

I'm drawing a blank on why that CSS var was actually used. 

As long as the color of the ResetLabel matches the "Saved" in the top toolbar when saving a draft post it will satisfy the request [here](https://github.com/WordPress/gutenberg/issues/44096#issuecomment-1247762119).

Further context and history can be found in https://github.com/WordPress/gutenberg/pull/44260.

**Comment by chad1008** [comment]
per @mirka's advice above, I've reverted all of the `@base-styles` and `@block-editor` overrides for now. ✅ 

> I would consider exposing a `COLORS.ui.themeDark10` that we can use here. (Aaand it looks like you already did that! 😄)

I hadn't actually carried that all the way into the exported `ui` object, but I have now! I've exposed (and applied) the new `COLORS.ui.themeDark10`, but we can use/not use that based on whether or not @aaronrobertshaw has any input re:

> Also unclear why this ResetLabel is a darker-10 and not just the theme color 🤔

**Comment by chad1008** [comment]
Sorry for the extra ping @aaronrobertshaw  - missed your reply above before!

File: packages/components/src/tools-panel/styles.ts (hunk 5 above)
Comment (chad1008) [reply packages/components/src/tools-panel/styles.ts]: Was looking at this with @mirka a little and this button doesn't have a disabled state, the 'reset' just isn't displayed when it isn't an option.

Looking at the focused/hovered state, I think I'd recommend we keep the current darker color. If the Reset span is given the same accent color as the rest of the list item, it becomes difficult to differentiate which button you're actually interacting with:

https://user-images.githubusercontent.com/13856531/199766554-d291cdf3-f9e8-4e53-a805-ff1d4f124b6f.mov

With the current darker color, there's at least a little contrast creating some distinction, though it _is_ really subtle. We may want something with more contrast/decoration, but probably not something for this PR.

Any objection to keeping the slightly darker reset for now, using the newly exposed variable?

Comment (mirka) [reply packages/components/src/tools-panel/styles.ts]: Yeah, we can keep it as is 👍 Thanks for looking into it.

**Review by mirka — APPROVED** [review]
Great work! 🚀

**Comment by strarsis** [comment]
That's nice! I had to override some editor styles because the editor UI uses the secondary color (which isn't always intended), this would also be changed by this PR?

**Comment by mirka** [comment]
> I had to override some editor styles because the editor UI uses the secondary color (which isn't always intended), this would also be changed by this PR?

@strarsis Hmm, I'm not sure what you mean by "secondary color" in this context. Could you open a bug report if you're still seeing the issue? If any UI colors are wrong, we'd like to fix them 🙂
